sktime/forecasting/compose/_multiplexer.py

```diff
@@ -84,83 +84,114 @@ class MultiplexForecaster(_HeterogenousEnsembleForecaster):
         "handles-missing-data": False,
         "scitype:y": "both",
         "y_inner_mtype": ["pd.DataFrame", "pd.Series"],
+        "fit_is_empty": False,
     }
 
+    _delegate_name = "forecaster_"
+
     def __init__(
         self,
         forecasters: list,
         selected_forecaster=None,
     ):
-        super(MultiplexForecaster, self).__init__(forecasters=forecasters, n_jobs=None)
+        super(MultiplexForecaster, self).__init__()
         self.selected_forecaster = selected_forecaster
 
+        self.forecasters = forecasters
+        self.forecasters_ = self._check_estimators(
+            forecasters,
+            attr_name="forecasters",
+            cls_type=BaseForecaster,
+            clone_ests=False,
+        )
+        self._set_forecaster()
+        self.clone_tags(self.forecaster_)
+        self.set_tags(**{"fit_is_empty": False})
+
```

Comment:
To clarify - not saying that this is wrong, just would like some clarity on the choice here.

Reply:
otherwise your test breaks!

That is in the case if the sequence of method calls is `fit`, `set_params`, `fit`.
If tag setting were not in `_fit`, they would not be correctly set.

This PR https://github.com/alan-turing-institute/sktime/pull/2531 would solve that, and also for any other estimator where tags are dynamically set, but until then we need it here or ... your test would break.

Reply:
hmm, so perhaps there is something I am missing here, but if I remove the two lines which coerce `fit_is_empty` to be `False` on my branch (which is merged with this branch) and run the pytests locally.... they actually seem to pass fine.  Which to me makes sense as all the forecasters we are testing with should have `fit_is_empty` being `False` and I believe that should get correctly copied over via `self.clone_tags(self.forecaster_)`.  Could you also try to see what I mean (just 2 lines to remove)?  

I guess the concern to me (and it is probably a small thing/weird edge case) is that this line could have weird behavior if for some reason someone was trying to use `MultiplexForecaster` with a forecaster whose `fit_is_empty` tag is `True`.  Not suggesting that someone should want to do such a thing - just don't know why the expected behavior in that case would be to set `fit_is_empty` to `False` manually.... That is more what I was asking.

Reply:
the problem arises under this condition, once you remove the lines in `_fit`:

* create the multiplexer any two forecasters
* use `set_params` to set the forecaster to forecaster1
* `fit` - this will produce results of forecaster2

reason: the tags to not get properly updated


It is even worse - let´s say we have `VAR` and `ThetaForecaster` - the former can only forecast on multivariate series, the latter only on univariate. Since `fit` errors out before `_fit` is reached, this doesn´t even work with the fix, and requires #2531 to work properly.

Reply:
so, for instance, this
```
from sktime.forecasting.theta import ThetaForecaster
from sktime.forecasting.var import VAR
from sktime.forecasting.compose import MultiplexForecaster
from sktime.utils._testing.series import _make_series

f = MultiplexForecaster([ThetaForecaster(), VAR()])
X = _make_series()
f.fit(X)
f.set_params(selected_forecaster="VAR")
X_multivariate = _make_series(n_columns=2)
f.fit(X_multivariate)
```
raises an error, even though it should not!

Reply:
Apologies I think I have perhaps not been totally clear what my issue is here, my issue is not with cloning the tags - it is with forcing the `fit_is_empty` tag to be `False`! 

I am with you - cloning the tags makes sense!  (since the fit() function is ultimately checking the data against `MultiplexForecaster`'s tags rather than it's `forecaster_` tags it makes sense that we should coerce them to be the same! 

I *think* I verified the hunch that `fit_is_empty` is not needed by slightly modifying your code and showing that cloning the tags is sufficient - no need to coerce `fit_is_empty` to be `False`. 

For example this does certainly break: 

```
f = MultiplexForecaster([("theta", ThetaForecaster()), ("var", VAR())])
X = _make_series()
f.fit(X)
f.selected_forecaster = "var"
f._set_forecaster()
X_multivariate = _make_series(n_columns=2)
f.fit(X_multivariate)
```

however I think adding the `clone_tags()` line alone rescues it: 

```
f = MultiplexForecaster([("theta", ThetaForecaster()), ("var", VAR())])
X = _make_series()
f.fit(X)
f.selected_forecaster = "var"
f._set_forecaster()
f.clone_tags(f.forecaster_)
X_multivariate = _make_series(n_columns=2)
f.fit(X_multivariate)
```

So really my confusion remains - I still don't understand why we need to coerce `fit_is_empty` to be `False`? (since it should be for all of the estimators we are cloning from - and if it is not for some reason - we should potentially raise some sort of error rather than coerce to being False?)   Perhaps it is a silly point, and I am missing something obvious though!!

*Also - would calling `clone_tags()` from within `_fit()` even solve this issue? (Reason I ask is that I understand `fit()` calls `_fit()` after it finishes, and some of the tests which would break from the tag issue (e.g. `_check_X_y()` would actually break before you ever got to the `clone_tags()` call in `_fit()`? *

Reply:
On a related note - I think the code example you provide above would make a great test case to add to the test suite! Just saying! :)

Reply:
Though in truth, reading #2458 I do think that would fix a lot of these issues on its own - so perhaps this conversation is not too helpful in the long run, if this is essentially just a temporary fix until that merges!

Reply:
> >I had meant change the two locations where fit_is_empty is coerced to be False
>
> Hm, not sure whether I understand your comment or what you are pointing at.

Ah, all I was trying to say here was that you call `self.set_tags(**{"fit_is_empty": False})` twice - once in the `__init__()` and once in `_fit()`.  And by change I meant remove (not change it to be false)

Reply:
> So as long as the tags in `MultiplexForecaster` are sufficiently generous to not block any inputs on the first pass, then the second pass will be able to check the input against the selected forecaster's more specific tags and raise the appropriate error then if neeeded? 

Hm, that's a good point. I did not think of this.
Indeed, we could let the inner forecaster's tags take care of the checking, conversion, vectorization.

*However*, I see a problem with retrieval and querying, so while tempting I fear this might not work.
The problem arises because the tags on the outside would no longer be "true", so they cannot be relied upon anymore for giving correct answers about properties.

For example, let's assume a user asking a pipeline "can you work with multivariate data"? The multiplexer would give the wrong answer.
It would also give the wrong answer if you further put the multiplexer in a pipeline, to the other parts of the pipeline that now wrap the multiplexer.
Another example, let's say we loop through forecaster instances in testing. We want to pass multivariate data to all forecasters that claim they are multivariate and check they do not raise the error. The multiplexer would then fail the test, because the inner forecaster cannot deal with the multivariate data, but the test would think that it should be able to.

Reply:
> Does this check out to you?

No, because in the test you would need to remove the `clone_tags` and the `_set_forecaster` line, these are not part of the public interface for the user.

Reply:
>>Does this check out to you?
>
>No, because in the test you would need to remove the clone_tags and the _set_forecaster line, these are not part >of the public interface for the user.
>Once you remove them, the franz test will fail.

Can in fact confirm it does.  Apologies, thought I had removed at least the `clone_tags()` line.  Guess I got too fast and loose with my examples!  (and perhaps fell into the fallacy of not being sufficiently skeptical of examples which conformed to my theory! 😱 )

>I'm calling it twice to deal with the situation you had in the test, where you call, in sequence, constructor, set_params, and fit. The fit_is_empty tag could be set incorrectly, in such a case.

Sure, seems reasonable.  I guess I was working on the assumption that `fit_is_empty` tag would be `False` for all the forecaster we wished to support in `MultiplexForecasters - and that was really the driving point of this whole line of questioning. But I do not have an issue with manually setting it per say/don't think it should cause any issues as is. 

Your point regarding public facing tags also seems reasonable to me. One could imagine a workaround, but this solution in conjunction with #2531 seems to work. 

Overall - if you are also on board, I think I am sufficiently satisfied for this thread to be marked as 'resolved' know it got a bit lengthy - thank you for putting the time in to follow up on these points with me!  Hopefully your other reviewers will find this (to some some small degree) useful as well!

Reply:
> Your point regarding public facing tags also seems reasonable to me. One could imagine a workaround, but this solution in conjunction with https://github.com/alan-turing-institute/sktime/pull/2531 seems to work.

Indeed - I've been thinking about a principled solution, and that's where #2531 came out of.
It seems to me now, if you try to think dynamic tags to its full extent, you need a reset or reset-like method.
Though I cannot prove mathematically or logically that this is necessarily the case.
Can you think of another principled solution which is not a hack or workaround? Or, do you see an argument?

> Overall - if you are also on board, I think I am sufficiently satisfied for this thread to be marked as 'resolved' know it got a bit lengthy - thank you for putting the time in to follow up on these points with me!

Well, thank you for thinking deeply about this, and uncovering the above issue!
I rarely get reviews that are this deep, precise, and to-the-point. Thank you.